pkg/endpoint/policy.go

```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
 	}
 }
 
-// Must be called with global endpoint.Mutex held
-func (e *Endpoint) regenerateConsumable(owner Owner) (bool, error) {
-	c := e.Consumable
+func getLabelsMap(owner Owner, cache *policy.ConsumableCache, minID, maxID policy.NumericIdentity) (labelArrays, error) {
+	labelsMap := map[policy.NumericIdentity]labels.LabelArray{}
+	reservedIDs := cache.GetReservedIDs()
+	var idx policy.NumericIdentity
+	for _, idx = range reservedIDs {
+		lbls, err := owner.GetCachedLabelList(idx)
+		if err != nil {
+			return nil, err
+		}
+		// Skip currently unused IDs
+		if lbls == nil || len(lbls) == 0 {
+			continue
+		}
+		labelsMap[idx] = lbls
+	}
+
+	for idx = minID; idx < maxID; idx++ {
+		lbls, err := owner.GetCachedLabelList(idx)
+		if err != nil {
+			return nil, err
+		}
+		// Skip currently unused IDs
+		if lbls == nil || len(lbls) == 0 {
+			continue
+		}
+		labelsMap[idx] = lbls
+	}
 
+	return labelsMap, nil
+}
+
+// Must be called with global endpoint.Mutex held
+func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap labelArrays, repo *policy.Repository, revision uint64, c *policy.Consumable) (bool, error) {
 	// Containers without a security label are not accessible
 	if c.ID == 0 {
 		log.Fatalf("[%s] BUG: Endpoints lacks identity", e.PolicyID())
 		return false, nil
 	}
 
-	cache := policy.GetConsumableCache()
-
-	// Skip if policy for this consumable is already valid
-	//if c.Iteration == cache.Iteration {
-	//	repo.Mutex.RUnlock()
-	//	log.Debugf("Reusing cached policy for identity %d", c.ID)
-	//	return false, nil
-	//}
-
-	maxID, err := owner.GetCachedMaxLabelID()
-	if err != nil {
-		return false, err
+	// Skip consumable recomputation for this consumable is already valid
+	if c.Iteration == revision {
+		log.Debugf("Reusing cached policy for identity %d", c.ID)
+		return false, nil
 	}
 
-	c.Mutex.RLock()
```

Comment:
I have since refactored this so that this should be clearer in the current version. The general problem is that we can't promote a read lock to a write lock without releasing the lock in between.

In this case (read) locking may have been mainly decorative as the consumable's labels should not change during the lifetime of the consumable, and the endpoint itself has a reference to it.

I've clarified the locking in the current revision of the series.